### ISSUE TITLE
httpproxy remove creds of proxy after processing request

### ISSUE THREAD

**Issue by NewUserHa** [scrapy/scrapy#4029]
it's hard to add different creds back to different proxy URL for reusing in next Request.

https://github.com/scrapy/scrapy/blob/master/scrapy/downloadermiddlewares/httpproxy.py#L53
this line should be removed.

**Comment by Gallaecio** [comment]
I suspect proxies will stop working if we remove that line?

Could you describe your use case in more detail, possibly with a [minimal code example](https://stackoverflow.com/help/mcve)? Maybe we can think of a different approach to handle your use case.

**Comment by NewUserHa** [comment]
tested and it'll not stop working.

eg:
a site needs a speficial proxy group and any of proxies fail need to be removed from the group. image that each proxy has different creds. with the line, once request had been downloaded, the creds were gone and no longer available to remove the exact proxy from the group.

**Comment by NewUserHa** [comment]
@kmike

**Comment by elacuesta** [comment]
Indeed, the mentioned line (https://github.com/scrapy/scrapy/blob/1.7.4/scrapy/downloadermiddlewares/httpproxy.py#L53) is necessary because the `proxy` meta value is then read by the downloader handler at https://github.com/scrapy/scrapy/blob/1.7.4/scrapy/core/downloader/handlers/http11.py#L283.
In any case, credentials should be popped after being processed there, although I'm not sure I understand the use case. Could you explain further?

**Comment by NewUserHa** [comment]
appearly it's not necessary as it can work without that line.
can you imagine you have three proxies:
a:a@a
b:b@a
c:c@a
removing that line is convenient for removing the bad proxy from proxy pool. meta is not good in nested request.

**Comment by Gallaecio** [comment]
If you are setting your proxies in a downloader middleware running before the proxy one, you can remove bad proxies there, right?
